File: 1439. Find the Kth Smallest Sum of a Matrix With Sorted Rows/Solution.cpp
#include <queue>
#include <tuple>
#include <vector>
using namespace std;

class Solution
{
public:
    int kthSmallest(vector<vector<int>> &mat, int k)
    {
        vector<int> buff{0};
        for (const vector<int> &row : mat)
            buff = kSmallestPairs(buff, row, k);
        return buff.back();
    }

private:
    vector<int> kSmallestPairs(const vector<int> &A, const vector<int> &B, size_t k)
    {
        vector<int> res;
        priority_queue<tuple<int, int, int>> pq;
        int m = A.size(), n = B.size();
        pq.emplace(-A[0] - B[0], 0, 0);
        while (!pq.empty() && res.size() < k)
        {
            auto [neg_sum, i, j] = pq.top();
            pq.pop();
            res.emplace_back(-neg_sum);
            if (j == 0 && i + 1 < m)
                pq.emplace(-A[i + 1] - B[j], i + 1, j);
            if (j + 1 < n)
                pq.emplace(-A[i] - B[j + 1], i, j + 1);
        }
        return res;
    }
};

File: 1439. Find the Kth Smallest Sum of a Matrix With Sorted Rows/Solution.py
from typing import List
import heapq


class Solution:
    def kthSmallest(self, mat: List[List[int]], k: int) -> int:
        def kSmallestPairs(A: List[int], B: List[int], k: int) -> List[int]:
            m, n = len(A), len(B)
            res = []
            heap = [(A[0] + B[0], 0, 0)]
            while heap and len(res) < k:
                total, i, j = heapq.heappop(heap)
                res.append(total)
                if j == 0 and i + 1 < m:
                    heapq.heappush(heap, (A[i + 1] + B[0], i + 1, 0))
                if j + 1 < n:
                    heapq.heappush(heap, (A[i] + B[j + 1], i, j + 1))
            return res

        buff = [0]
        for row in mat:
            buff = kSmallestPairs(buff, row, k)
        return buff[-1]
